modules/network/src/test/java/org/apache/ignite/internal/network/serialization/ClassDescriptorParserTest.java

```diff
@@ -0,0 +1,355 @@
+        public ExternalizableClass() {
+        }
+
+        @Override
+        public void writeExternal(ObjectOutput out) throws IOException {
+            // No-op.
+        }
+
+        @Override
+        public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+            // No-op.
+        }
+    }
+
+    @Test
+    public void testExternalizable() {
+        ClassDescriptor descriptor = parser.parse(ExternalizableClass.class);
+
+        assertEquals(SerializationType.EXTERNALIZABLE, descriptor.serializationType());
+        assertEquals(0, descriptor.fields().size());
+    }
+
+    static class ArbitraryClass {
+    }
+
+    @Test
+    public void testArbitrary() {
+        ClassDescriptor descriptor = parser.parse(ArbitraryClass.class);
+
+        assertEquals(SerializationType.ARBITRARY, descriptor.serializationType());
+        assertEquals(0, descriptor.fields().size());
+    }
+
+    @SuppressWarnings({"ExternalizableWithoutPublicNoArgConstructor", "RedundantThrows"})
+    static class ExternalizableWithoutPublicNoArgConstructor implements Externalizable {
+        public ExternalizableWithoutPublicNoArgConstructor(@SuppressWarnings("unused") int unused) {
+        }
+
+        @Override
+        public void writeExternal(ObjectOutput out) throws IOException {
+            // No-op.
+        }
+
+        @Override
+        public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+            // No-op.
+        }
+    }
+
+    @Test
+    public void testExternalizableWithoutConstructor() {
+        assertThrows(IgniteException.class,
+                () -> parser.parse(ExternalizableWithoutPublicNoArgConstructor.class));
+    }
+
+    /** Class that holds a set of all types parseable by the class descriptor parser. */
+    @SuppressWarnings("unused")
+    static class Holder {
+        private SerializableClass field1;
+
+        private ExternalizableClass field2;
+
+        private ArbitraryClass field3;
+
+        private int primitiveField;
+    }
+
+    @Test
+    public void testHolderClass() {
+        ClassDescriptor holderDescriptor = parser.parse(Holder.class);
+
+        ClassDescriptor serializableDescriptor = context.getDescriptor(SerializableClass.class);
+        assertNotNull(serializableDescriptor);
+
+        ClassDescriptor externalizableDescriptor = context.getDescriptor(ExternalizableClass.class);
+        assertNotNull(externalizableDescriptor);
+
+        ClassDescriptor arbitraryDescriptor = context.getDescriptor(ArbitraryClass.class);
+        assertNotNull(arbitraryDescriptor);
+
+        ClassDescriptor intDescriptor = context.getDescriptor(DefaultType.INT.descriptorId());
+        assertNotNull(intDescriptor);
+
+        List<FieldDescriptor> fields = holderDescriptor.fields();
+        assertEquals(4, fields.size());
+
+        Map<String, FieldDescriptor> fieldsMap = fields.stream()
```

Comment:
Why do you need a map here? I think using an array is enough

Reply:
Why is field index not enough?

Reply:
Excerpt from the javadocs of the Class#getDeclaredFields:
```
The elements in the returned array are not sorted and are not in any particular order.
```

Reply:
oh. ok then